lib/bolt/cli.rb

```diff
@@ -385,681 +350,309 @@ def handle_parser_errors
-                                  inventory_version: inventory.version)
-      end
+      Bolt::Application.new(
+        analytics:     analytics,
+        config:        config,
+        executor:      executor,
+        inventory:     inventory,
+        log_outputter: log_outputter,
+        outputter:     outputter,
+        pal:           pal,
+        plugins:       plugins,
+        rerun:         rerun
+      )
+    rescue Bolt::Error => e
+      outputter.fatal_error(e)
+      raise e
+    end
 
-      analytics.screen_view(screen, **screen_view_fields)
+    # Execute a Bolt command. The +options+ hash includes the subcommand and
+    # action to be run, as well as any additional arguments and options for the
+    # command.
+    #
+    # @param options [Hash] The CLI options.
+    #
+    def execute(options)
+      config = load_config(options)
 
-      case options[:action]
-      when 'show'
-        case options[:subcommand]
-        when 'task'
-          if options[:object]
-            show_task(options[:object])
-          else
-            list_tasks
-          end
-        when 'plan'
-          if options[:object]
-            show_plan(options[:object])
-          else
-            list_plans
-          end
-        when 'inventory'
-          if options[:detail]
-            show_targets
-          else
-            list_targets
-          end
-        when 'group'
-          list_groups
-        when 'module'
-          list_modules
-        when 'plugin'
-          list_plugins
-        end
-        return 0
-      when 'convert'
-        pal.convert_plan(options[:object])
-        return 0
+      analytics = begin
+        client = Bolt::Analytics.build_client(config.analytics)
+        client.bundled_content = bundled_content(options)
+        client
       end
 
-      message = 'There might be processes left executing on some nodes.'
+      outputter = Bolt::Outputter.for_format(
+        config.format,
+        config.color,
+        options[:verbose],
+        config.trace,
+        config.spinner
+      )
+
+      app = load_application(analytics, config, outputter, options)
 
-      if %w[task plan].include?(options[:subcommand]) && options[:task_options] && !options[:params_parsed] && pal
-        options[:task_options] = pal.parse_params(options[:subcommand], options[:object], options[:task_options])
+      handler = Signal.trap :INT do |signo|
+        Bolt::Logger.logger(self).info(
+          "Exiting after receiving SIG#{Signal.signame(signo)} signal. There might be processes left "\
+          "executing on some targets."
+        )
+        exit!
       end
 
       case options[:subcommand]
-      when 'guide'
-        code = if options[:object]
-                 show_guide(options[:object])
-               else
-                 list_topics
-               end
-      when 'project'
+      when 'apply'
+        app.apply(options[:object], options)
```

Comment:
This is definitely something for a later PR, but having separate applications for each subcommand that get instantiated here rather than before the case statement might make more sense than having one big application that can everything. That'll keep the classes relatively small, and potentially avoid loading things we don't have to for some subcommands.